lib/pure/net.nim

```diff
@@ -678,56 +672,49 @@ when defineSsl:
-    copyMem(psk, pskString.cstring, pskString.len)
-
-    return pskString.len.cuint
-
-  proc `clientGetPskFunc=`*(ctx: SslContext, fun: SslClientGetPskFunc) =
-    ## Sets function that returns the client identity and the PSK based on identity
-    ## hint from the server.
-    ##
-    ## Only used in PSK ciphersuites.
-    ctx.getExtraInternal().clientGetPskFunc = fun
-    ctx.context.SSL_CTX_set_psk_client_callback(
-        if fun == nil: nil else: pskClientCallback)
-
-  proc serverGetPskFunc*(ctx: SslContext): SslServerGetPskFunc =
-    return ctx.getExtraInternal().serverGetPskFunc
+  template genpskServerCallback(pskfunc: SslServerGetPskFunc): auto =
+    proc pskServerCallback(ssl: SslCtx; identity: cstring; psk: ptr cuchar;
+        max_psk_len: cint): cuint {.cdecl.} =
+      let pskString = pskfunc($identity)
+      if pskString.len.cint > max_psk_len:
+        return 0
+      copyMem(psk, pskString.cstring, pskString.len)
+      return pskString.len.cuint
 
-  proc pskServerCallback(ssl: SslCtx; identity: cstring; psk: ptr cuchar;
-      max_psk_len: cint): cuint {.cdecl.} =
-    let ctx = SslContext(context: ssl.SSL_get_SSL_CTX)
-    let pskString = (ctx.serverGetPskFunc)($identity)
-    if pskString.len.cint > max_psk_len:
-      return 0
-    copyMem(psk, pskString.cstring, pskString.len)
+    pskServerCallback
 
-    return pskString.len.cuint
-
-  proc `serverGetPskFunc=`*(ctx: SslContext, fun: SslServerGetPskFunc) =
+  proc `serverGetPskFunc=`*(ctx: SslContext, fun: static SslServerGetPskFunc) =
     ## Sets function that returns PSK based on the client identity.
     ##
     ## Only used in PSK ciphersuites.
-    ctx.getExtraInternal().serverGetPskFunc = fun
-    ctx.context.SSL_CTX_set_psk_server_callback(if fun == nil: nil
-                                                else: pskServerCallback)
+    if not fun.isNil:
+      ctx.context.SSL_CTX_set_psk_server_callback(genpskServerCallback(fun))
+
+  template genpskClientCallback(pskfunc: SslClientGetPskFunc): auto =
+      proc pskClientCallback(ssl: SslPtr; hint: cstring; identity: cstring;
+          max_identity_len: cuint; psk: ptr cuchar;
+          max_psk_len: cuint): cuint {.cdecl.} =
+        let (identityString, pskString) = pskfunc($hint)
+        if pskString.len.cuint > max_psk_len:
+          return 0
+        if identityString.len.cuint >= max_identity_len:
+          return 0
+        copyMem(identity, identityString.cstring, identityString.len + 1) # with the last zero byte
+        copyMem(psk, pskString.cstring, pskString.len)
+        return pskString.len.cuint
+
+      pskClientCallback
+
+  proc `clientGetPskFunc=`*(ctx: SslContext, fun: static SslClientGetPskFunc) =
+    ## Sets function that returns the client identity and the PSK based on identity
+    ## hint from the server.
+    ##
+    ## Only used in PSK ciphersuites.
+    if not fun.isNil:
+      ctx.context.SSL_CTX_set_psk_client_callback(genpskClientCallback(fun))
```

Comment:
Shouldn't we remove the callback when `nil` is passed in? These were the semantics previously.

Reply:
great catch. will do

Reply:
importantly, if the server callback is nil, the server will accept _any_ psk sent from the client.